module/zcommon/zfs_fletcher.c

```diff
@@ -222,30 +264,219 @@ fletcher_4_incremental_native(const void *buf, uint64_t size,
 	ZIO_SET_CHECKSUM(zcp, a, b, c, d);
 }
 
+int
+fletcher_4_impl_set(const char *val)
+{
+	const fletcher_4_ops_t *ops;
+	enum fletcher_selector idx;
+	size_t val_len;
+	unsigned i;
+
+	val_len = strlen(val);
+	while ((val_len > 0) && !!isspace(val[val_len-1])) /* trim '\n' */
+		val_len--;
+
+	for (i = 0; i < ARRAY_SIZE(fletcher_4_impl_selectors); i++) {
+		const char *name = fletcher_4_impl_selectors[i].fis_name;
+
+		if (val_len == strlen(name) &&
+		    strncmp(val, name, val_len) == 0) {
+			idx = i;
+			break;
+		}
+	}
+	if (i >= ARRAY_SIZE(fletcher_4_impl_selectors))
+		return (-EINVAL);
+
+	ops = fletcher_4_impl_selectors[idx].fis_ops;
+	if (ops == NULL || (ops->valid != NULL && !ops->valid()))
+		return (-ENOTSUP);
+
+	mutex_enter(&fletcher_4_impl_lock);
+	if (fletcher_4_impl_chosen != idx)
+		fletcher_4_impl_chosen = idx;
+	mutex_exit(&fletcher_4_impl_lock);
+
+	return (0);
+}
+
+static inline const fletcher_4_ops_t *
+fletcher_4_impl_get(void)
+{
+#if !defined(_KERNEL)
+	if (fletcher_4_impl_chosen == FLETCHER_CYCLE) {
+		static volatile unsigned int cycle_count = 0;
+		const fletcher_4_ops_t *ops = NULL;
+		unsigned int index;
+
+		while (1) {
+			index = atomic_inc_uint_nv(&cycle_count);
+			ops = fletcher_4_algos[
+			    index % ARRAY_SIZE(fletcher_4_algos)];
+			if (ops->valid == NULL || ops->valid())
```

Comment:
Is this right?  Shouldn't it be checking that the selected algorithm is valid before breaking out of the loop:
`if (ops->valid != NULL && ops->valid())`